datasource/src/test/scala/quasar/physical/mongo/MongoSpec.scala

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright 2014–2018 SlamData Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package quasar.physical.mongo
+
+import slamdata.Predef._
+import quasar.physical.mongo.MongoResource.{Collection, Database}
+
+import cats.syntax.apply._
+import cats.effect.IO
+import fs2.Stream
+import org.specs2.mutable.Specification
+import org.specs2.execute.AsResult
+import org.mongodb.scala._
+import org.bson.{Document => _, _}
+import scala.io.Source
+import shims._
+import testImplicits._
+
+class MongoSpec extends Specification {
+  import MongoSpec._
+
+
+  step(MongoSpec.setupDB)
+
+  "can create client from valid connection string" >> {
+    mkMongo.attempt.compile.last.unsafeRunSync() match {
+      case None => AsResult(false).updateMessage("Impossible happened, Mongo.apply.attempt must return something")
+      case Some(Left(thr)) => AsResult(false).updateMessage("Mongo.apply returned an error" ++ thr.getMessage())
+      case Some(Right(_)) => AsResult(true)
+    }
+  }
+
+  "getting databases works correctly" >> {
+    val stream = for {
+      mongo <- mkMongo
+      databases <- mongo.databases
+    } yield databases
+    val evaluatedDbs = stream.compile.toList.unsafeRunSync()
+    val expectedDbs = MongoSpec.dbs.map(Database(_))
+    expectedDbs.toSet.subsetOf(evaluatedDbs.toSet)
+  }
+
+  "databaseExists returns true for existing dbs" >>  {
+    val stream = for {
+      mongo <- mkMongo
+      db <- MongoSpec.correctDbs
+      exists <- mongo.databaseExists(db)
+    } yield exists
+    stream.fold(true)(_ && _).compile.last.unsafeRunSync().getOrElse(false)
+  }
```

Comment:
Instead of using 1 test in combination with `stream.fold` I'd split these into multiple tests which are easier to check for correctness, like this:

```
    MongoSpec.dbs.foreach { db =>
      s"databaseExists returns true for existing db $db" >> {
        ...
      }
    }
```

(or you can even just take 1 db)

Similar for tests below

Reply:
Ok